### ISSUE TITLE
Feature Idea: consul_kv allow json import to key value store

### ISSUE THREAD

**Issue by Etherdaemon** [ansible/ansible-modules-extras#2104]
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

consul_kv
##### SUMMARY

Currently the consul_kv module written by @sgargan doesn't really support an import function of a json blob. I think it would be really useful to implement this so that you can import/fill a consul key value store and have the json file that is human readable and parseable by other json products.
##### PROPOSAL

Just wondering if there are any objections to updating the module to perform the following:
- take a json variable and check if the json input is set during the execute method
- change key,value variable to required false
- insert methods to read in the json file or input, convert the dictionary to be consul readable (keys to full paths)
- call the existing add_value method inside a iterating loop to add the key/values to consul
- update the existing add_value method to not exit_json straight after a value is added but instead pass back the relevant values to the calling methods

I'm hesitating on doing this at the moment as although I've managed to get it to work for my own implementation, it required quite a bit of adjustment to the module due to where the exit points were originally as well as the type of returned data.
##### INPUT JSON

```
{
  "app_a": {
    "auto_commit_interval_ms": "10000",
    "batch_num_messages": "75",
    "compression_codec": "gzip",
    "fetch_min_bytes": "8192",
    "fetch_wait_max_ms": "300",
    "message_send_max_retries": "10",
    "producer_type": "async",
    "request_required_acks": "-1",
    "retry_backoff_ms": "500"
  },
  "app_b": {
    "log4j_appender_FA_BufferedIO": {
        "testing_nested": "value_test"
    },
    "log4j_appender_FA_DatePattern": ".yyyy-MM-dd-HH",
    "log4j_appender_FA_DateRollEnforced": "true",
    "log4j_appender_FA_File": "/etc/test.txt",
    "log4j_appender_FA_MaxFileSize": "10GB"
  }
}
```
##### OUTPUT PARSEABLE BY CONSUL

```
{   u'app_a/auto_commit_interval_ms': u'10000',
    u'app_a/batch_num_messages': u'75',
    u'app_a/compression_codec': u'gzip',
    u'app_a/fetch_min_bytes': u'8192',
    u'app_a/fetch_wait_max_ms': u'300',
    u'app_a/message_send_max_retries': u'10',
    u'app_a/producer_type': u'async',
    u'app_a/request_required_acks': u'-1',
    u'app_a/retry_backoff_ms': u'500',
    u'app_b/log4j_appender_FA_BufferedIO/testing_nested': u'value_test',
    u'app_b/log4j_appender_FA_DatePattern': u'.yyyy-MM-dd-HH',
    u'app_b/log4j_appender_FA_DateRollEnforced': u'true',
    u'app_b/log4j_appender_FA_File': u'/etc/test.txt',
    u'app_b/log4j_appender_FA_MaxFileSize': u'10GB'}
```

**Comment by sgargan** [comment]
What is your intended access pattern? Do you have a use case that requires the overhead of a k/v pair per document value. Without reason to lock or version at this fine a grain you'd be much better off storing the document for each app and retrieving it in one go, If your primary usecase is for configuration this is definitely the case. You're unlikely to need to version and lock 'log4j_appender_FA_File' independently of 'log4j_appender_FA_BufferedIO'. If they don't need to be versioned independently they should go in the same document. Perhaps you could explain the use case here and what problem you are trying to solve.

**Comment by Etherdaemon** [comment]
Hi @sgargan,

I was using the above more as an example of a json blob.

But in terms of use case, for myself personally, I need to import environment configurations into the key/value store and have said environment configs in a human readable format that can be used by both consul as well as being able to be parsed by other apps/languages such as Ruby and Python. When I provision an environment I would like to import into the key/value store the relevant configs - EG application related configs, db names, etc etc.

Currently the python consulate lib reckons it can export to json but when you try it, it spits out an array of arrays full of non-readable key/values and also not parseable by by other languages in your normal json parsing ways as its not technically json. Also its import of json is also not of a normal json file either which sucks.

With your existing module, you support adding in individual key values but to do a full import I would need to somehow pop it into a variables file as a dictionary for ansible to loop through which makes it a little difficult when you are working with nested dictionaries with dynamic data. It made sense in my scenario to read in a json file with the relevant data and iterate over it to import into the store.

Does that elaborate enough? Hopefully the above made sense.

**Comment by sgargan** [comment]
I get what you're trying to do but not why you're trying to it. Each key
value pair should be something you could potentially lock around, the unit
of write being the group of elements you'd like to be written and accessed
atomically. Writing each element of a json blob is going to hurt your
performance and if you're going to end up reading them sequentially or even
in groups is going to end up being wasted work. I would rethink your
usecase and question if this is really a good idea.

**Comment by ansibot** [comment]
@sgargan, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sgargan** [comment]
Going to close this for the time being @Etherdaemon feel free to comment if you'd like to discuss further. marking 'wontfix' for ansibot
